rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/security/ResourceAuthenticationProvider.java

```diff
@@ -0,0 +1,114 @@
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.crypto.bcrypt.BCrypt;
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
+import org.springframework.stereotype.Component;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.eclipse.sw360.rest.resourceserver.Sw360ResourceServer.API_TOKEN_HASH_SALT;
+
+@Profile("!SECURITY_MOCK")
+@Component
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class ResourceAuthenticationProvider implements AuthenticationProvider {
+
+    private static final Logger log = Logger.getLogger(ResourceAuthenticationProvider.class);
+
+    @NotNull
+    private final Sw360UserService userService;
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        if (authentication.isAuthenticated()) {
+            log.debug("Authentication already authenticated");
+            return authentication;
+        }
+
+        String token = authentication.getCredentials().toString();
+        String tokenHash = BCrypt.hashpw(token, API_TOKEN_HASH_SALT);
+        User sw360User = getUserFromTokenHash(tokenHash);
+        RestApiToken userToken = getApiTokenFromUser(tokenHash, sw360User);
+        log.debug("Valid token authentication for user: " + sw360User.getEmail());
+        Set<GrantedAuthority> authorities = new HashSet<>();
+        userToken.getAuthorities().stream().forEach(a -> authorities.add(new SimpleGrantedAuthority(a)));
+        Authentication preAuthenticatedToken = new PreAuthenticatedAuthenticationToken(sw360User.getEmail(), token, authorities);
+        preAuthenticatedToken.setAuthenticated(true);
+        return preAuthenticatedToken;
+    }
+
+    private User getUserFromTokenHash(String tokenHash) {
+        try {
+            return userService.getUserByApiToken(tokenHash);
+        } catch (RuntimeException e) {
+            log.debug("Could not find any user for the entered token");
+            throw new AuthenticationServiceException("Your entered API token is not valid.");
+        }
+    }
+
+    private RestApiToken getApiTokenFromUser(String tokenHash, User sw360User) {
+        Optional<RestApiToken> apiToken = sw360User.getRestApiTokens().stream().
+                filter(t -> t.getToken().equals(tokenHash)).findFirst();
+        if (apiToken.isPresent() && !isApiTokenExpired(apiToken.get())) {
+            return apiToken.get();
+        } else {
+            log.debug("Could not load API token form user " + sw360User.getEmail());
```

Comment:
the log message does not match the exception below. `"Could not load API token..."` does not match invalid API token.

Reply:
changed handling of log/exception messages